src/Interpreters/MonotonicityCheckVisitor.h

```diff
@@ -125,13 +147,55 @@ class MonotonicityCheckMatcher
         }
 
         ColumnsWithTypeAndName args;
-        args.emplace_back(data.arg_data_type, "tmp");
+        ColumnWithTypeAndName const_arg;
+
+        if (arguments_size == 2)
+        {
+            if (ast_function.arguments->children[0]->as<ASTLiteral>())
+            {
+                const auto * literal = ast_function.arguments->children[0]->as<ASTLiteral>();
+                const_arg = extractLiteralColumnAndTypeFromAstLiteral(literal);
+                args.push_back(const_arg);
+                args.emplace_back(data.arg_data_type, "tmp");
+            }
+            else
+            {
+                const auto * literal = ast_function.arguments->children[1]->as<ASTLiteral>();
+                args.emplace_back(data.arg_data_type, "tmp");
+                const_arg = extractLiteralColumnAndTypeFromAstLiteral(literal);
+                args.push_back(const_arg);
+            }
+        }
+        else
+        {
+            args.emplace_back(data.arg_data_type, "tmp");
+        }
+
         auto function_base = function->build(args);
 
         if (function_base && function_base->hasInformationAboutMonotonicity())
         {
             bool is_positive = data.monotonicity.is_positive;
-            data.monotonicity = function_base->getMonotonicityForRange(*data.arg_data_type, Field(), Field());
+            data.monotonicity = function_base->getMonotonicityForRange(*data.arg_data_type, data.range.left, data.range.right);
+
+            auto & key_range = data.range;
+
+            /// If we apply function to open interval, we can get empty intervals in result.
+            /// E.g. for ('2020-01-03', '2020-01-20') after applying 'toYYYYMM' we will get ('202001', '202001').
+            /// To avoid this we make range left and right included.
+            /// Any function that treats NULL specially is not monotonic.
+            /// Thus we can safely use isNull() as an -Inf/+Inf indicator here.
+            if (!key_range.left.isNull())
+            {
+                key_range.left = applyFunction(function_base, data.arg_data_type, key_range.left);
+                key_range.left_included = true;
+            }
+
+            if (!key_range.right.isNull())
+            {
+                key_range.right = applyFunction(function_base, data.arg_data_type, key_range.right);
+                key_range.right_included = true;
+            }
```

Comment:
`data.range` (same as `key_range`) is overridden before `visit` invocation:
``` cpp
MonotonicityCheckVisitor::Data data {{table_with_columns}, destination_storage.getContext(), {}};

data.range = hyperrectangle[i];

MonotonicityCheckVisitor(data).visit(expression_list->children[i]);
```

Is the updated value used at all?

Reply:
It should. If I am not mistaking, `MonotonicityCheckVisitor` is called once per expression in the partition expression list.

A single expression might be contain several stacked-up functions (e.g, `toInt(toFloat('3.14'))`). The caller (`MergeTreePartitionCompatibilityVerifier`) assigns the inner most range to be checked. In this case, something like `3.14` and `3.15`.

`MonotonicityCheckVisitor` should then visit each function recursively and update the range accordingly. I did not check the visit logic, but I assume it is correct.